consoles/sshVirtsh.pm

```diff
@@ -325,111 +326,126 @@ sub add_interface {
     return;
 }
 
-sub add_disk {
-    my ($self, $args) = @_;
+sub to_byte {
+    my $sp              = shift;
+    my $size_multiplier = {
+        b => 1,
+        k => 1024, K => 1024,
+        M => 1024 * 1024,
+        G => 1024 * 1024 * 1024,
+        T => 1024 * 1024 * 1024 * 1024,
+    };
 
-    my $backingfile             = $args->{backingfile};
-    my $cdrom                   = $args->{cdrom};
-    my $name                    = $self->name;
-    my $file                    = $name . $args->{dev_id} . ($self->vmm_family eq 'vmware' ? '.vmdk' : '.img');
-    my $basedir                 = '/var/lib/libvirt/images/';
-    my $vmware_datastore        = get_var('VMWARE_DATASTORE', '');
-    my $vmware_openqa_datastore = "/vmfs/volumes/$vmware_datastore/openQA/";
-    if ($args->{create}) {
-        my $size = $args->{size} || '20G';
-        if ($self->vmm_family eq 'vmware') {
-            my $vmware_disk_path = $vmware_openqa_datastore . $file;
-            # Power VM off, delete it's disk image, and create it again.
-            # Than wait for some time for the VM to *really* turn off.
-            my $cmd =
-              "( set -x; vmid=\$(vim-cmd vmsvc/getallvms | awk \'/$name/ { print \$1 }\');" .
-              'if [ $vmid ]; then ' .
-              'vim-cmd vmsvc/power.off $vmid;' .
-              'vim-cmd vmsvc/destroy $vmid;' .
-              'fi;' .
-              "vmkfstools -v1 -U $vmware_disk_path;" .
-              "vmkfstools -v1 -c $size --diskformat thin $vmware_disk_path; sleep 10 ) 2>&1";
-            my $retval = $self->run_cmd($cmd, domain => 'sshVMwareServer');
-            die "Can't create VMware image $vmware_disk_path" if $retval;
+    #in case of invalid format return error
+    if ($sp !~ /^(?<num>\d+)(?<unit>[bkKMGT])?$/) {
+        return undef;
+    } else {
+        return $+{num} * $size_multiplier->{$+{unit} // 'b'};
+    }
+}
+
+sub gen_kvm_img {
+    my ($self, $args) = @_;
+    # create [--object OBJECTDEF] [-q] [-f FMT] [-b BACKING_FILE] [-F BACKING_FMT] [-u] [-o OPTIONS] FILENAME [SIZE]
```

Comment:
I would add input validation like:
```suggestion
    die('Missing argument size or invalid format') unless(defined(to_byte($args->{size})));
    die('Missing argument `new_image_file`) unless(defined($args->{new_image_file}));
    $args->{backingfile} //= 0; # default is __no__ backing file
    
    # create [--object OBJECTDEF] [-q] [-f FMT] [-b BACKING_FILE] [-F BACKING_FMT] [-u] [-o OPTIONS] FILENAME [SIZE]
```

Reply:
Isn't the mandatory only new file parameter ?

Reply:
Didn't saw a place where you set a default size in `get_kvm_img()`. And the line also check if `to_byte()` can parse the input.